math/cosine.go

```diff
@@ -117,61 +105,68 @@ func getCosineP(negXSq *decimal.Big) func(n uint64) *decimal.Big {
 	}
 }
 
-func getCosineB() func(n uint64) *decimal.Big {
-	return func(n uint64) *decimal.Big {
-		return one
-	}
-}
-
-func getCosineQ(precision int) func(n uint64) *decimal.Big {
+func getCosineQ(ctx decimal.Context) func(n uint64) *decimal.Big {
+	var q, tmp decimal.Big
 	return func(n uint64) *decimal.Big {
-		//(0) = 1, q(n) = 2n(2n-1) for n > 0
+		// (0) = 1, q(n) = 2n(2n-1) for n > 0
 		if n == 0 {
 			return one
 		}
 
-		//most of the time n will be a small number so
+		// most of the time n will be a small number so
 		// use the fastest method to calculate 2n(2n-1)
+		const cosine4NMaxN = 2147483648
 		if n < cosine4NMaxN {
-			return new(decimal.Big).SetUint64(((n * n) << 2) - (n << 1))
+			return q.SetUint64((2 * n) * (2*n - 1)) // ((n*n) << 2) - (n << 1)
```

Comment:
((n*n) << 2) - (n << 1) has one less multiply in it and should be faster

Reply:
I don't believe the compiler will make the leap to this.

Reply:
Heh, nope. The compiler wins again. The compiler notices that `2*n` is computed twice, so it computes it once using a shift, then multiplies the result by itself less one. The compiler doesn't know what the shifting version does, so it just executes it verbatim. :-)

Reply:
That is awesome, I don't believe the GCC would have done that well at all (or at least the one that I used last which I think was v5). At any rate, keep the more readable one.